types/node/querystring.d.ts

```diff
@@ -10,7 +10,11 @@ declare module "querystring" {
 
     interface ParsedUrlQuery { [key: string]: string | string[]; }
 
-    function stringify(obj?: {}, sep?: string, eq?: string, options?: StringifyOptions): string;
+    // When these types support TypeScript 3.0+, we can replace this with `unknown`.
+    type PoorMansUnknown = {} | null | undefined;
+    interface ParsedUrlQueryInput { [key: string]: PoorMansUnknown; }
```

Comment:
Alternatively we could do `{ [key: string]: string | string[] | number | number[] | boolean | boolean[]; }`, since these are all the types `querystring.stringify` and `url.format` will implicitly convert to a `string`:

``` sh-session
$ node
> require('querystring').stringify({ string: 'string', number: 1, boolean: true, numbers: [1,2], booleans: [true, false], strings: ['string1', 'string2'] })
'string=string&number=1&boolean=true&numbers=1&numbers=2&booleans=true&booleans=false&strings=string1&strings=string2'
```

IIUC, other value types like objects will be omitted:

``` sh-session
$ node
> require('querystring').stringify({ object: { foo: 1 } })
'object='
```

Currently with `unknown` we allow these types to be passed in—after all it doesn't generate a runtime error, so it is "safe".

https://devdocs.io/node/querystring#querystring_querystring_stringify_obj_sep_eq_options

Reply:
I agree here, the `PoorMansUnknown` is a huge hack (as suggested by the name itself), we should wait for node typings to move beyond TS 3.x instead.

Reply:
By agreeing, are you saying to replace `PoorMansUnknown` with `string | string[] | number | number[] | boolean | boolean[]`?

FWIW, I don't think it's a huge hack—there are no downsides to using it, other than the lack of a name. Even TypeScript itself uses it when it can't infer a generic.

I think we should consider the desired behaviour and then use that to pick between the two options here.

Reply:
Someone may a "query builder class" which works with `PoorMansUnknown` but not with `string | string[] | number | number[] | boolean | boolean[]`:
```
class MyQuery {
    val: string;
    constructor(a: number) {
         this.val = `number${a}`;
    }
}
```